lib/thread.c

```diff
@@ -108,6 +107,7 @@ M0_INTERNAL void *m0_thread_trampoline(void *arg)
 	M0_PRE(t->t_tls.tls_m0_instance != NULL);
 	M0_PRE(t->t_tls.tls_self == t);
 
+	t->t_tls.tls_loci = m0_processor_id_get();
```

Comment:
This locality-related initialisation code is pertinent to the thread (each thread has its own `tls_loci` value). And it is used by any thread calling `m0_locality_here()`, not only by locality thread. So it seems like an appropriate place to me.

Also, we do reset this value at the "locality initialisation function", see `loc_thr_init() -> m0_thread_confine()`.

Reply:
> And it is used by any thread calling m0_locality_here(), not only by locality thread.

Then it could be zeroed here and set to the appropriate value at the appropriate place.

Reply:
Sorry, I don't understand - why? Which "appropriate place"?

Reply:
@andriytk: something like this:

```
> git diff
diff --git a/fop/fom.c b/fop/fom.c
index 3db79333..93e56d89 100644
--- a/fop/fom.c
+++ b/fop/fom.c
@@ -970,6 +970,7 @@ static void loc_handler_thread(struct m0_loc_thread *th)
  */
 static int loc_thr_init(struct m0_loc_thread *th)
 {
+       th->lt_thread.t_tls.tls_loci = th->lt_loc->fl_idx;
        return m0_thread_confine(&th->lt_thread, &th->lt_loc->fl_processors);
 }
 
diff --git a/lib/thread.c b/lib/thread.c
index 3eaea41d..3caef762 100644
--- a/lib/thread.c
+++ b/lib/thread.c
@@ -107,7 +107,6 @@ M0_INTERNAL void *m0_thread_trampoline(void *arg)
        M0_PRE(t->t_tls.tls_m0_instance != NULL);
        M0_PRE(t->t_tls.tls_self == t);
 
-       t->t_tls.tls_loci = m0_processor_id_get();
        m0_set(t->t_tls.tls_m0_instance);
        m0_addb2_global_thread_enter();
        if (t->t_init != NULL) {
@@ -140,12 +139,6 @@ M0_INTERNAL void m0_thread_shun(void)
 M0_INTERNAL int m0_thread_confine(struct m0_thread *q,
                                  const struct m0_bitmap *processors)
 {
-       int cpu_idx = m0_bitmap_ffs(processors);
-
-       M0_PRE(cpu_idx >= 0);
-
-       q->t_tls.tls_loci = cpu_idx;
-
        return m0_thread_arch_confine(q, processors);
 }
 ```
The comment for `m0_thread_tls::tls_loci` should also mention that this is locality index for locality threads and 0 for the rest of the threads.

Reply:
This will mean that all non-locality threads will use the same locality, which might increase contention over some resources, like AST-queue, and impact performance. So I'd prefer to not do this.